protocol/proto/addr_enc_dec.go

```diff
@@ -0,0 +1,254 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+const (
+	defaultRetryForever = true
+)
+
+var (
+	errNoValidConnection = errors.New("no valid connection available")
+)
+
+type connectFn func(addr string) (net.Conn, error)
+
+type addrEncdec struct {
+	sync.RWMutex
+
+	encdec              ConnectionEncodeDecoder
+	addr                string
+	validConn           *atomic.Bool // Use atomic for r/w fast path.
+	resetCh             chan struct{}
+	resetNanos          int64
+	retrier             retry.Retrier
+	dialTimeout         time.Duration
+	reconnectDelayNanos int64
+	doneCh              chan struct{}
+	closed              bool
+	logger              log.Logger
+	m                   *addrEncdecMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+type addrEncdecMetrics struct {
+	resetConn      tally.Counter
+	resetConnError tally.Counter
+}
+
+// NewAddressEncodeDecoder creates an EncodeDecoder that
+// manages the connection to the target address.
+func NewAddressEncodeDecoder(
+	addr string,
+	opts AddressEncodeDecoderOptions,
+) AddressEncodeDecoder {
+	if opts == nil {
+		opts = NewAddressEncodeDecoderOptions()
+	}
+	c := addrEncdec{
+		encdec:     NewConnectionEncodeDecoder(nil, opts.ConnectionEncodeDecoderOptions()),
+		addr:       addr,
+		validConn:  atomic.NewBool(false),
+		resetCh:    make(chan struct{}, 1),
+		resetNanos: time.Now().UnixNano(),
+		// Should retry forever on reconnect.
+		retrier:             retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		dialTimeout:         opts.DialTimeout(),
+		reconnectDelayNanos: opts.ReconnectDelay().Nanoseconds(),
+		doneCh:              make(chan struct{}),
+		closed:              false,
+		logger:              opts.InstrumentOptions().Logger(),
+		m:                   newAddrEncdecMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:               time.Now,
+	}
+
+	c.connectFn = c.connectOnce
+	if err := c.resetConn(c.connectFn); err != nil {
+		c.signalReset()
+	}
+	return &c
+}
+
+func newAddrEncdecMetrics(m tally.Scope) *addrEncdecMetrics {
+	return &addrEncdecMetrics{
+		resetConn:      m.Counter("reset-conn"),
+		resetConnError: m.Counter("reset-conn-error"),
+	}
+}
+
+func (c *addrEncdec) Encode(msg Marshaler) error {
+	if !c.hasValidConnection() {
+		return errNoValidConnection
+	}
+	if err := c.encdec.Encode(msg); err != nil {
+		c.signalInvalidConnection()
+		return err
+	}
+	return nil
+}
+
+func (c *addrEncdec) Decode(acks Unmarshaler) error {
```

Comment:
Do you need to take a read lock during decoding? Otherwise there's a race condition where you might still decode after the object is closed.

Reply:
Possibly, can put this in a read lock